url147ce07ebe_JoshuaKissoon_Kademlia_tgz-pJ8-kademlia_simulations_RoutingTableStateTestingJ8/src/kademlia/KadServer.java

```diff
@@ -0,0 +1,357 @@
+                                if (task != null)
+                                {
+                                    task.cancel();
+                                }
+                            }
+                        }
+                        else
+                        {
+                            /* There is currently no receivers, try to get one */
+                            receiver = messageFactory.createReceiver(messCode, this);
+                        }
+
+                        /* Invoke the receiver */
+                        if (receiver != null)
+                        {
+                            receiver.receive(msg, comm);
+                        }
+                    }
+                }
+                catch (IOException e)
+                {
+                    //this.isRunning = false;
+                    System.err.println("Server ran into a problem in listener method. Message: " + e.getMessage());
+                }
+            }
+        }
+        finally
+        {
+            if (!socket.isClosed())
+            {
+                socket.close();
+            }
+            this.isRunning = false;
+        }
+    }
+
+    /**
+     * Remove a conversation receiver
+     *
+     * @param comm The id of this conversation
+     */
+    private synchronized void unregister(int comm)
+    {
+        receivers.remove(comm);
+        this.tasks.remove(comm);
+    }
+
+    /**
+     * Stops listening and shuts down the server
+     */
+    public synchronized void shutdown()
+    {
+        this.isRunning = false;
+        this.socket.close();
+        timer.cancel();
+    }
+
+    /**
+     * Task that gets called by a separate thread if a timeout for a receiver occurs.
+     * When a reply arrives this task must be canceled using the <code>cancel()</code>
+     * method inherited from <code>TimerTask</code>. In this case the caller is
+     * responsible for removing the task from the <code>tasks</code> map.
+     * */
+    class TimeoutTask extends TimerTask
+    {
+
+        private final int comm;
+        private final Receiver recv;
+
+        public TimeoutTask(int comm, Receiver recv)
+        {
+            this.comm = comm;
+            this.recv = recv;
+        }
+
+        @Override
+        public void run()
+        {
+            if (!KadServer.this.isRunning)
+            {
+                return;
+            }
+
+            try
+            {
+                unregister(comm);
+                recv.timeout(comm);
+            }
+            catch (IOException e)
+            {
+                System.err.println("Cannot unregister a receiver. Message: " + e.getMessage());
+            }
+        }
+    }
+
+    public void printReceivers()
+    {
+        for (Integer r : this.receivers.keySet())
+        {
+            System.out.println("Receiver for comm: " + r + "; Receiver: " + this.receivers.get(r));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.